text/0000-cargo-test-coverage.md

```diff
@@ -0,0 +1,343 @@
+a Rust developer would have to either update the `RUSTFLAGS` environment variable or Cargo
+manifest keys with `-C instrument-coverage`. This would automatically enable instrumentation
+of all Rust crates within the dependency graph, not just the current crate. Instrumenting
+all crates including transitive dependencies does not help the developer ensure test coverage
+for the crates they actually want to test. This also adds unnecessary work for both the Rust
+compiler and codegen backend, `LLVM` in this case, to instrument all libraries as opposed to
+the subset a Rust developer actually cares about. This support is currently limited to the
+`LLVM` codegen backend and will have no effect when another codegen backend is used.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This section examines the features proposed by this RFC:
+
+## CLI option
+
+Three new flags will be added for the `cargo test` command, `--coverage`, `--coverage-format` and `--coverage-output-directory`
+The `--coverage` flag would instruct Cargo to add the rustc flag `-C instrument-coverage`, for the current
+crate only. Cargo also supports two types of workspaces, a workspace with a root package and one without called a
+virtual manifest. In the case where the `cargo test --coverage` command is run from a Cargo workspace which has a root package,
+the `-C instrument-coverage` rustc flag would only be enabled for the root package. In the case where the `cargo test --coverage`
+command is run from a Cargo workspace without a root package, the `-C instrument-coverage` rustc flag would be enabled for
+all default members of the workspace. This would mean that only the crates selected would be instrumented and code
+coverage results would only be collected for those crates and not all crates in the dependency graph.
+
+As an example, let's take the following crate `foo`:
+
+```text
+/Cargo.toml
+/src
+    +-- lib.rs
+```
+
+Where crate `foo` has a dependency on the `regex`:
+```toml
+[dependencies]
+regex = "*"
+```
+
+And `lib.rs` contains:
+
+```rust
+use regex::*;
+
+pub fn match_regex(re: Regex, text: &str) -> bool {
+    let Some(caps) = re.captures(text) else {
+        return false
+    };
+
+    let matches = caps.iter().filter_map(|s| s).collect::<Vec<Match>>();
+    matches.len() > 0
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn find_match() {
+        let result = match_regex(Regex::new(".*").unwrap(), "Hello World");
+        assert_eq!(result, true);
+    }
+
+    #[test]
+    fn find_no_match() {
+        let result = match_regex(Regex::new("a+").unwrap(), "Hello World");
+        assert_eq!(result, false);
+    }
+}
+```
+
+Now running `cargo test --coverage` would build the `foo` crate and instrument all libraries created by this crate.
+Each test executable run will generate a unique `*.profraw` file. At the end of all test execution, Cargo will be
```

Comment:
> However, I lean towards doctests being a blocker. User expectations will be that `--coverage` reports everything and will be confused that doctests are not included.

I think that makes the most since from an end user perspective. I'll update the RFC to include doctests in this RFC as a core feature as opposed to in the `Future Possibilities` section. Thanks!